lib/prolongator.in.cpp

```diff
@@ -1,50 +1,118 @@
 #include "multigrid.h"
+#include <blas_quda.h>
 
 namespace quda
 {
 
   template <int...> struct IntList {
   };
 
-  template <int fineColor, int coarseColor, int... N>
+  template <int fineColor, int coarseColor, int nVec, int... N>
+  void ProlongateMma2(ColorSpinorField &out, const ColorSpinorField &in, const ColorSpinorField &v,
+      const int *fine_to_coarse, const int *const *spin_map, int parity, IntList<nVec, N...>) {
+    if (out.Nvec() == nVec) {
+      ProlongateMma<fineColor, coarseColor, nVec>(out, in, v, fine_to_coarse, spin_map, parity);
+    } else {
+      if constexpr (sizeof...(N) > 0) {
+        ProlongateMma2<fineColor, coarseColor>(out, in, v, fine_to_coarse, spin_map, parity, IntList<N...>());
+      } else {
+        errorQuda("nVec = %d has not been instantiated", out.Nvec());
+      }
+    }
+  }
+
+  template <class F> auto create_color_spinor_copy(cvector_ref<F> &fs, QudaFieldOrder order)
+  {
+    ColorSpinorParam param(fs[0]);
+    int nVec = (fs.size() + 7) / 8 * 8; // Make a multiple of 8
+    param.nColor = fs[0].Ncolor() * nVec;
+    param.nVec = nVec;
+    param.create = QUDA_NULL_FIELD_CREATE;
+    param.fieldOrder = order;
+    return ColorSpinorField(param);
+  }
+
+  static auto create_color_spinor_copy(const ColorSpinorField &f, QudaFieldOrder order)
+  {
+    ColorSpinorParam param(f);
+    param.create = QUDA_NULL_FIELD_CREATE;
+    param.fieldOrder = order;
+    return ColorSpinorField(param);
+  }
+
+  template <bool use_mma, int fineColor, int coarseColor, int... N>
   void Prolongate2(cvector_ref<ColorSpinorField> &out, cvector_ref<const ColorSpinorField> &in, const ColorSpinorField &v,
                    const int *fine_to_coarse, const int *const *spin_map, int parity, IntList<coarseColor, N...>)
   {
     if (in[0].Ncolor() == coarseColor) {
       if constexpr (coarseColor >= fineColor) {
-        Prolongate<fineColor, coarseColor>(out, in, v, fine_to_coarse, spin_map, parity);
+        if constexpr (use_mma) {
+
+          constexpr QudaFieldOrder csOrder = QUDA_SPACE_SPIN_COLOR_FIELD_ORDER;
+          ColorSpinorField v_in = create_color_spinor_copy(in, csOrder);
+          ColorSpinorField v_out = create_color_spinor_copy(out, csOrder);
+          ColorSpinorField V = create_color_spinor_copy(v, csOrder);
+          BlockTransposeForward(v_in, in);
+          V.copy(v);
+
+          IntList<@QUDA_MULTIGRID_MRHS_LIST@> nvecs;
+          ProlongateMma2<fineColor, coarseColor>(v_out, v_in, V, fine_to_coarse, spin_map, parity, nvecs);
+
+          bool to_non_rel = (out.Nspin() == 4) && (out[0].GammaBasis() == QUDA_UKQCD_GAMMA_BASIS);
+          BlockTransposeBackward(v_out, out, to_non_rel);
+#if 0
```

Comment:
Is there a reason to keep this around or can it be deleted?

Reply:
These are essentially the only test code for prolonator and restrictor (with/without MMA) - we might want to have a separate test/ctest for prolongator and restrictor.

Reply:
https://github.com/lattice/quda/pull/1497/commits/1be04f52b1f513c9e5a522a759884f736329e487